app/scripts/directives/mobileServiceInstanceRow.js

```diff
@@ -0,0 +1,125 @@
+'use strict';
+
+(function() {
+  angular.module('openshiftConsole').component('mobileServiceInstanceRow', {
+    controller: [
+      '$filter',
+      '$routeParams',
+      '$scope',
+      'APIService',
+      'AuthorizationService',
+      'BindingService',
+      'DataService',
+      'ListRowUtils',
+      MobileServiceInstanceRow
+    ],
+    controllerAs: 'row',
+    bindings: {
+        apiObject: '<',
+        serviceClass: '<',
+        state: '<'
+    },
+    templateUrl: 'views/_mobile-service-instance-row.html'
+  });
+
+  function MobileServiceInstanceRow($filter,
+                              $routeParams,
+                              $scope,
+                              APIService,
+                              AuthorizationService,
+                              BindingService,
+                              DataService,
+                              ListRowUtils) {
+    var row = this;
+    var serviceInstanceDisplayName = $filter('serviceInstanceDisplayName');
+    row.serviceBindingsVersion = APIService.getPreferredVersion('servicebindings');
+    row.state = row.state || {};
+    var annotationSpace = "org.aerogear.binding." + this.apiObject.metadata.name;
+    var extendedAnnotationSpace = "org.aerogear.binding-ext." + this.apiObject.metadata.name;
+    var watches = [];
+    var bindingsWatch = false;
+    row.bindingsLimit = _.get(row.serviceClass, "spec.externalMetadata.bindingsLimit", 1);
+    _.extend(row, ListRowUtils.ui);
+
+    $scope.$on('$destroy', function(){
+        DataService.unwatchAll(watches);
+        DataService.unwatch(bindingsWatch);
+    });
+
+    watches.push(DataService.watchObject(APIService.getPreferredVersion('mobileclients'), $routeParams.mobileclient, {namespace: $routeParams.project}, function(mobileClient) {
```

Comment:
I think this should all really be called form `$onInit`, rather than be loosely scattered in the `MobileServiceInstanceRow` function body.  Preferably broken down into named functions.

Reply:
:+1:

Reply:
The reason this was not in `$onInit` was that we extend `ListRowUtils.ui` which has it's own `$onInit`.

```.js
 _.extend(row, ListRowUtils.ui);
```

I have enclosed this in `$onChanges` guarded by `isFirstChange`
```.js
    row.$onChanges = function(changes) {
      row.displayName = serviceInstanceDisplayName(row.apiObject, row.serviceClass);

      if (changes.apiObject && changes.apiObject.isFirstChange()) {
        row.getClient()
          .then(function(client) {
            row.mobileClient = client;
            row.serviceType = _.get(row, 'serviceClass.spec.externalMetadata.serviceName');

            row.sortAnnotations();
            row.getBindingParametersData();
            row.watchServiceBindings();
            row.watchClient();
          });       
      }
    };
```